SDK crashes app after a while in case the tracking backend is permanently not reachable

We are currently using the Piwik SDK in one client app. Last week the staging environment was undergoing maintenance and was not reachable for a few days / returning HTTP 500 error codes for every call made by the Piwik SDK.

It seems that the SDK then retries the calls without a retry limit until the app crashes with a SSLException "Unable to create application data", which seems to indicate a resource exhaustion due to too many threads/files etc. (At least according to [this](https://github.com/square/okhttp/issues/3761) issue in the OkHttp repo). This behaviour was logged with the version 2.0.0 of the Piwik SDK integrated, but could also be reproduced with the 3.0.0 version of the Piwik SDK.

In the described case, the SDK always retries the calls without a retry limit, indicated by the endlessly repeating logs like this:

```
04-04 17:23:19.183 18907-18946/? D/PIWIK:EventCache: Switched state to ONLINE, uncached 0 events from disk.
04-04 17:23:19.184 18907-18946/? D/PIWIK:Dispatcher: Drained 1 events.
04-04 17:23:19.351 18907-18946/? D/PIWIK:DefaultPacketSender: status code 500
04-04 17:23:19.358 18907-18946/? D/PIWIK:Dispatcher: Unsuccesful assuming OFFLINE, requeuing events.
04-04 17:23:19.360 18907-18946/? D/PIWIK:Dispatcher: Dispatched 0 events.
04-04 17:23:19.364 18907-18946/? D/PIWIK:EventDiskCache: Cache check took 0ms
04-04 17:23:19.365 18907-18946/? D/PIWIK:EventDiskCache: Uncaching of 0 events took 0ms
04-04 17:23:19.365 18907-18946/? D/PIWIK:EventCache: Switched state to ONLINE, uncached 0 events from disk.
```

After a while multiple SocketExceptions appear in the logs:
```
04-04 13:44:47.387 30360-30405/? D/PIWIK:EventCache: Switched state to ONLINE, uncached 0 events from disk.
04-04 13:44:47.387 30360-30405/? D/PIWIK:Dispatcher: Drained 2 events.
04-04 13:44:47.389 30360-30405/? D/PIWIK:Dispatcher: java.net.SocketException: socket already closed
        at java.net.PlainSocketImpl.socketClose0(PlainSocketImpl.java:211)
        at java.net.AbstractPlainSocketImpl.socketClose(AbstractPlainSocketImpl.java:721)
        at java.net.AbstractPlainSocketImpl.close(AbstractPlainSocketImpl.java:546)
        at java.net.SocksSocketImpl.close(SocksSocketImpl.java:554)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:375)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:200)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:182)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:357)
        at java.net.Socket.connect(Socket.java:616)
        at com.android.okhttp.internal.Platform.connectSocket(Platform.java:145)
        at com.android.okhttp.internal.io.RealConnection.connectSocket(RealConnection.java:141)
        at com.android.okhttp.internal.io.RealConnection.connect(RealConnection.java:112)
        at com.android.okhttp.internal.http.StreamAllocation.findConnection(StreamAllocation.java:184)
        at com.android.okhttp.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:126)
        at com.android.okhttp.internal.http.StreamAllocation.newStream(StreamAllocation.java:95)
        at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:281)
        at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:224)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:461)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:127)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getOutputStream(HttpURLConnectionImpl.java:258)
        at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getOutputStream(DelegatingHttpsURLConnection.java:218)
        at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getOutputStream(Unknown Source:0)
        at org.piwik.sdk.dispatcher.Dispatcher.dispatch(Dispatcher.java:247)
        at org.piwik.sdk.dispatcher.Dispatcher$1.run(Dispatcher.java:167)
        at java.lang.Thread.run(Thread.java:764)
    Unsuccesful assuming OFFLINE, requeuing events.
    Dispatched 0 events.
04-04 13:44:47.391 30360-30405/? D/PIWIK:EventDiskCache: Cache check took 0ms
    Uncaching of 0 events took 0ms
```

After a few minutes, the app then crashes with the following exception, preceded by "AppData::create pipe (2) failed: Too many open files":
```
04-04 14:56:58.513 5634-5697/? E/NativeCrypto: AppData::create pipe(2) failed: Too many open files
04-04 14:56:58.532 5634-5697/? E/AndroidRuntime: FATAL EXCEPTION: Thread-5
    Process: ?, PID: 5634
    java.lang.RuntimeException: javax.net.ssl.SSLException: Unable to create application data
        at com.android.org.conscrypt.ConscryptFileDescriptorSocket.newSsl(ConscryptFileDescriptorSocket.java:161)
        at com.android.org.conscrypt.ConscryptFileDescriptorSocket.<init>(ConscryptFileDescriptorSocket.java:152)
        at com.android.org.conscrypt.OpenSSLSocketFactoryImpl.createSocket(OpenSSLSocketFactoryImpl.java:149)
        at com.android.okhttp.internal.io.RealConnection.connectTls(RealConnection.java:181)
        at com.android.okhttp.internal.io.RealConnection.connectSocket(RealConnection.java:149)
        at com.android.okhttp.internal.io.RealConnection.connect(RealConnection.java:112)
        at com.android.okhttp.internal.http.StreamAllocation.findConnection(StreamAllocation.java:184)
        at com.android.okhttp.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:126)
        at com.android.okhttp.internal.http.StreamAllocation.newStream(StreamAllocation.java:95)
        at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:281)
        at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:224)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:461)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:127)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getOutputStream(HttpURLConnectionImpl.java:258)
        at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getOutputStream(DelegatingHttpsURLConnection.java:218)
        at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getOutputStream(Unknown Source:0)
        at org.piwik.sdk.dispatcher.Dispatcher.dispatch(Dispatcher.java:247)
        at org.piwik.sdk.dispatcher.Dispatcher$1.run(Dispatcher.java:167)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: javax.net.ssl.SSLException: Unable to create application data
        at com.android.org.conscrypt.NativeCrypto.SSL_new(Native Method)
        at com.android.org.conscrypt.SslWrapper.newInstance(SslWrapper.java:58)
        at com.android.org.conscrypt.ConscryptFileDescriptorSocket.newSsl(ConscryptFileDescriptorSocket.java:159)
        at com.android.org.conscrypt.ConscryptFileDescriptorSocket.<init>(ConscryptFileDescriptorSocket.java:152) 
        at com.android.org.conscrypt.OpenSSLSocketFactoryImpl.createSocket(OpenSSLSocketFactoryImpl.java:149) 
        at com.android.okhttp.internal.io.RealConnection.connectTls(RealConnection.java:181) 
        at com.android.okhttp.internal.io.RealConnection.connectSocket(RealConnection.java:149) 
        at com.android.okhttp.internal.io.RealConnection.connect(RealConnection.java:112) 
        at com.android.okhttp.internal.http.StreamAllocation.findConnection(StreamAllocation.java:184) 
        at com.android.okhttp.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:126) 
        at com.android.okhttp.internal.http.StreamAllocation.newStream(StreamAllocation.java:95) 
        at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:281) 
        at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:224) 
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:461) 
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:127) 
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getOutputStream(HttpURLConnectionImpl.java:258) 
        at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getOutputStream(DelegatingHttpsURLConnection.java:218) 
        at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getOutputStream(Unknown Source:0) 
        at org.piwik.sdk.dispatcher.Dispatcher.dispatch(Dispatcher.java:247) 
        at org.piwik.sdk.dispatcher.Dispatcher$1.run(Dispatcher.java:167) 
        at java.lang.Thread.run(Thread.java:764) 
```

--- a/piwik-sdk/src/main/java/org/piwik/sdk/dispatcher/DefaultPacketSender.java
+++ b/piwik-sdk/src/main/java/org/piwik/sdk/dispatcher/DefaultPacketSender.java
@@ -2,9 +2,11 @@ package org.piwik.sdk.dispatcher;
 
 import org.piwik.sdk.Piwik;
 
+import java.io.BufferedReader;
 import java.io.BufferedWriter;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.InputStreamReader;
 import java.io.OutputStream;
 import java.io.OutputStreamWriter;
 import java.net.HttpURLConnection;
@@ -23,6 +25,7 @@ public class DefaultPacketSender implements PacketSender {
         HttpURLConnection urlConnection = null;
         try {
             urlConnection = (HttpURLConnection) packet.getTargetURL().openConnection();
+            Timber.tag(LOGGER_TAG).v("Connection open to %s", urlConnection.getURL().toExternalForm());
             urlConnection.setConnectTimeout((int) mTimeout);
             urlConnection.setReadTimeout((int) mTimeout);
 
@@ -44,9 +47,10 @@ public class DefaultPacketSender implements PacketSender {
                         gzipStream.write(toPost.getBytes(Charset.forName("UTF8")));
                     } finally {
                         // If closing fails we assume the written data to be invalid.
-                        // Don't catch the exception and let it abort the `send` call.
+                        // Don't catch the exception and let it abort the `send(Packet)` call.
                         if (gzipStream != null) gzipStream.close();
                     }
+
                     OutputStream outputStream = null;
                     try {
                         outputStream = urlConnection.getOutputStream();
@@ -86,11 +90,33 @@ public class DefaultPacketSender implements PacketSender {
             }
 
             int statusCode = urlConnection.getResponseCode();
-            Timber.tag(LOGGER_TAG).d("status code %s", statusCode);
+            final boolean successful = checkResponseCode(statusCode);
 
-            return checkResponseCode(statusCode);
+            if (successful) {
+                Timber.tag(LOGGER_TAG).v("Transmission succesful (code=%d).", statusCode);
+            } else {
+                // Consume the error stream (or at least close it) if the statuscode was non-OK (not 2XX)
+                final StringBuilder errorReason = new StringBuilder();
+                BufferedReader errorReader = null;
+                try {
+                    errorReader = new BufferedReader(new InputStreamReader(urlConnection.getErrorStream()));
+                    String line;
+                    while ((line = errorReader.readLine()) != null) errorReason.append(line);
+                } finally {
+                    if (errorReader != null) {
+                        try {
+                            errorReader.close();
+                        } catch (IOException e) {
+                            Timber.tag(LOGGER_TAG).d(e, "Failed to close the error stream.");
+                        }
+                    }
+                }
+                Timber.tag(LOGGER_TAG).w("Transmission failed (code=%d, reason=%s)", statusCode, errorReason.toString());
+            }
+
+            return successful;
         } catch (Exception e) {
-            Timber.tag(LOGGER_TAG).e(e, "Sending failed");
+            Timber.tag(LOGGER_TAG).e(e, "Transmission failed unexpectedly.");
             return false;
         } finally {
             if (urlConnection != null) urlConnection.disconnect();

--- a/piwik-sdk/src/main/java/org/piwik/sdk/dispatcher/DefaultPacketSender.java
+++ b/piwik-sdk/src/main/java/org/piwik/sdk/dispatcher/DefaultPacketSender.java
@@ -4,6 +4,8 @@ import org.piwik.sdk.Piwik;
 
 import java.io.BufferedWriter;
 import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
 import java.io.OutputStreamWriter;
 import java.net.HttpURLConnection;
 import java.nio.charset.Charset;
@@ -32,6 +34,7 @@ public class DefaultPacketSender implements PacketSender {
 
                 final String toPost = packet.getPostData().toString();
                 if (mGzip) {
+
                     urlConnection.addRequestProperty("Content-Encoding", "gzip");
                     ByteArrayOutputStream byteArrayOS = new ByteArrayOutputStream();
 
@@ -39,15 +42,43 @@ public class DefaultPacketSender implements PacketSender {
                     try {
                         gzipStream = new GZIPOutputStream(byteArrayOS);
                         gzipStream.write(toPost.getBytes(Charset.forName("UTF8")));
-                    } finally { if (gzipStream != null) gzipStream.close();}
+                    } finally {
+                        // If closing fails we assume the written data to be invalid.
+                        // Don't catch the exception and let it abort the `send` call.
+                        if (gzipStream != null) gzipStream.close();
+                    }
+                    OutputStream outputStream = null;
+                    try {
+                        outputStream = urlConnection.getOutputStream();
+                        outputStream.write(byteArrayOS.toByteArray());
+                    } finally {
+                        if (outputStream != null) {
+                            try {
+                                outputStream.close();
+                            } catch (IOException e) {
+                                // Failing to close the stream is not enough to consider the transmission faulty.
+                                Timber.tag(LOGGER_TAG).d(e, "Failed to close output stream after writing gzipped POST data.");
+                            }
+                        }
+                    }
 
-                    urlConnection.getOutputStream().write(byteArrayOS.toByteArray());
                 } else {
+
                     BufferedWriter writer = null;
                     try {
                         writer = new BufferedWriter(new OutputStreamWriter(urlConnection.getOutputStream(), "UTF-8"));
                         writer.write(toPost);
-                    } finally { if (writer != null) writer.close(); }
+                    } finally {
+                        if (writer != null) {
+                            try {
+                                writer.close();
+                            } catch (IOException e) {
+                                // Failing to close the stream is not enough to consider the transmission faulty.
+                                Timber.tag(LOGGER_TAG).d(e, "Failed to close output stream after writing POST data.");
+                            }
+                        }
+                    }
+
                 }
 
             } else { // GET

